File: apps/fsharp/src/MachineLearning/Chapter15/RequestValidation.fs
/// プレゼンテーション層。HTTP の本文（JSON）を検証して、ドメイン層の型にする
module MachineLearning.Chapter15.RequestValidation

open System.Text.Json
open MachineLearning.Chapter15.Domain

/// 検証の結果。正しければ値を、不正なら理由の一覧を持つ
type Validation<'T> = Result<'T, string list>

/// 項目ごとの検証をまとめる計算式。and! でつないだ項目は、どれかが不正でも残りを検証し、理由をすべて集める
type ValidationBuilder() =
    member _.BindReturn(result: Validation<'T>, f: 'T -> 'U) : Validation<'U> = Result.map f result

    member _.MergeSources(left: Validation<'T>, right: Validation<'U>) : Validation<'T * 'U> =
        match left, right with
        | Ok l, Ok r -> Ok(l, r)
        | Error l, Error r -> Error(l @ r)
        | Error errors, Ok _
        | Ok _, Error errors -> Error errors

let validation = ValidationBuilder()

/// 項目の値。無い項目と null は None
let private tryField (name: string) (json: JsonElement) : JsonElement option =
    match json.TryGetProperty name with
    | true, value when value.ValueKind <> JsonValueKind.Null -> Some value
    | _ -> None

let private required (name: string) (value: JsonElement option) : Validation<JsonElement> =
    value |> Option.map Ok |> Option.defaultValue (Error [ $"{name} を指定してください" ])

let private number (name: string) (value: JsonElement) : Validation<float> =
    if value.ValueKind = JsonValueKind.Number then
        Ok(value.GetDouble())
    else
        Error [ $"{name} は数値にしてください" ]

let private integer (name: string) (value: JsonElement) : Validation<int> =
    match value.ValueKind, value.TryGetInt32() with
    | JsonValueKind.Number, (true, n) -> Ok n
    | _ -> Error [ $"{name} は整数にしてください" ]

/// inline にすると、呼び出す場所ごとに float・int など実際の型で展開される（GenericZero はその型の 0）
let inline private notNegative (name: string) (value: 'N) : Validation<'N> =
    if value >= LanguagePrimitives.GenericZero then
        Ok value
    else
        Error [ $"{name} は 0 以上にしてください" ]

/// JSON の値（数値か文字列）を、決められた値の一覧のどれかに対応づける
let private oneOf (name: string) (choices: (string * 'T) list) (value: JsonElement) : Validation<'T> =
    let raw =
        match value.ValueKind with
        | JsonValueKind.String -> value.GetString()
        | _ -> value.GetRawText()

    match choices |> List.tryFind (fun (key, _) -> key = raw) with
    | Some(_, choice) -> Ok choice
    | None ->
        let keys = choices |> List.map fst |> String.concat "、"
        Error [ $"{name} は {keys} のどれかにしてください" ]

let private nonNegativeNumber name json =
    tryField name json
    |> required name
    |> Result.bind (number name)
    |> Result.bind (notNegative name)

let private nonNegativeInteger name json =
    tryField name json
    |> required name
    |> Result.bind (integer name)
    |> Result.bind (notNegative name)

/// 本文を JSON として読み、オブジェクトなら検証の関数に渡す
let private parseWith (validate: JsonElement -> Validation<'T>) (body: string) : Validation<'T> =
    try
        use document = JsonDocument.Parse body

        if document.RootElement.ValueKind = JsonValueKind.Object then
            validate document.RootElement
        else
            Error [ "JSON のオブジェクトにしてください" ]
    with :? JsonException ->
        Error [ "JSON の形式が正しくありません" ]

let parseMovie (body: string) : Validation<Movie> =
    body
    |> parseWith (fun json ->
        validation {
            let! sns1 = nonNegativeNumber "sns1" json
            and! sns2 = nonNegativeNumber "sns2" json
            and! actor = nonNegativeNumber "actor" json

            and! original =
                tryField "original" json
                |> required "original"
                |> Result.bind (oneOf "original" [ "0", false; "1", true ])

            return
                {
                    Sns1 = sns1
                    Sns2 = sns2
                    Actor = actor
                    Original = original
                }
        })

/// 省略できる項目。無ければ None、あれば検証する
let private optional (name: string) (validate: JsonElement -> Validation<'T>) (json: JsonElement) =
    match tryField name json with
    | None -> Ok None
    | Some value -> validate value |> Result.map Some

let parsePassenger (body: string) : Validation<Passenger> =
    body
    |> parseWith (fun json ->
        validation {
            let! pclass =
                tryField "pclass" json
                |> required "pclass"
                |> Result.bind (oneOf "pclass" [ "1", First; "2", Second; "3", Third ])

            and! sex =
                tryField "sex" json
                |> required "sex"
                |> Result.bind (oneOf "sex" [ "male", Male; "female", Female ])

            and! age =
                json
                |> optional "age" (fun value -> number "age" value |> Result.bind (notNegative "age"))

            and! sibSp = nonNegativeInteger "sib_sp" json
            and! parch = nonNegativeInteger "parch" json
            and! fare = nonNegativeNumber "fare" json

            and! embarked =
                json
                |> optional "embarked" (oneOf "embarked" [ "C", Cherbourg; "Q", Queenstown; "S", Southampton ])

            return
                {
                    Pclass = pclass
                    Sex = sex
                    Age = age
                    SibSp = sibSp
                    Parch = parch
                    Fare = fare
                    Embarked = embarked
                }
        })

File: apps/csharp/src/MachineLearning/Chapter15/RequestValidation.cs
namespace MachineLearning.Chapter15;

using System.Text.Json;

/// <summary>
/// プレゼンテーション層。HTTP の本文（JSON）を検証して、ドメイン層の型にする。
/// 不正な項目が複数あれば、理由をすべて集めて返す。
/// </summary>
public static class RequestValidation
{
    /// <summary>検証の結果。正しければ値を、不正なら理由の一覧を持つ。</summary>
    public sealed record Validation<T>(T? Value, IReadOnlyList<string> Errors)
    {
        public bool IsValid => this.Errors.Count == 0;
    }

    /// <summary>本文を映画の特徴量にする。</summary>
    public static Validation<Movie> ParseMovie(string body) =>
        ParseWith(body, json =>
        {
            var errors = new List<string>();
            var sns1 = NonNegativeNumber(json, "sns1", errors);
            var sns2 = NonNegativeNumber(json, "sns2", errors);
            var actor = NonNegativeNumber(json, "actor", errors);
            var original = OneOf(json, "original", new Dictionary<string, bool>(StringComparer.Ordinal) { ["0"] = false, ["1"] = true }, errors);
            return errors.Count == 0
                ? Valid(new Movie(sns1!.Value, sns2!.Value, actor!.Value, original!.Value))
                : Invalid<Movie>(errors);
        });

    /// <summary>本文を乗客の特徴量にする。年齢と乗船港は省略できる。</summary>
    public static Validation<Passenger> ParsePassenger(string body) =>
        ParseWith(body, json =>
        {
            var errors = new List<string>();
            var pclass = OneOf(json, "pclass", new Dictionary<string, Pclass>(StringComparer.Ordinal)
            {
                ["1"] = Pclass.First,
                ["2"] = Pclass.Second,
                ["3"] = Pclass.Third,
            }, errors);
            var sex = OneOf(json, "sex", new Dictionary<string, Sex>(StringComparer.Ordinal)
            {
                ["male"] = Sex.Male,
                ["female"] = Sex.Female,
            }, errors);
            var age = OptionalNonNegativeNumber(json, "age", errors);
            var sibSp = NonNegativeInteger(json, "sib_sp", errors);
            var parch = NonNegativeInteger(json, "parch", errors);
            var fare = NonNegativeNumber(json, "fare", errors);
            var embarked = OptionalOneOf(json, "embarked", new Dictionary<string, Embarked>(StringComparer.Ordinal)
            {
                ["C"] = Embarked.Cherbourg,
                ["Q"] = Embarked.Queenstown,
                ["S"] = Embarked.Southampton,
            }, errors);
            return errors.Count == 0
                ? Valid(new Passenger(pclass!.Value, sex!.Value, age, sibSp!.Value, parch!.Value, fare!.Value, embarked))
                : Invalid<Passenger>(errors);
        });

    private static Validation<T> Valid<T>(T value) => new(value, []);

    private static Validation<T> Invalid<T>(IReadOnlyList<string> errors) => new(default, errors);

    private static Validation<T> ParseWith<T>(string body, Func<JsonElement, Validation<T>> validate)
    {
        try
        {
            using var document = JsonDocument.Parse(body);
            return document.RootElement.ValueKind == JsonValueKind.Object
                ? validate(document.RootElement)
                : Invalid<T>(["JSON のオブジェクトにしてください"]);
        }
        catch (JsonException)
        {
            return Invalid<T>(["JSON の形式が正しくありません"]);
        }
    }

    /// <summary>項目の値。無い項目と null は null を返す。</summary>
    private static JsonElement? Field(JsonElement json, string name) =>
        json.TryGetProperty(name, out var value) && value.ValueKind != JsonValueKind.Null ? value : null;

    private static double? NonNegativeNumber(JsonElement json, string name, List<string> errors)
    {
        var value = Field(json, name);
        if (value is null)
        {
            errors.Add($"{name} を指定してください");
            return null;
        }

        return NonNegative(name, Number(name, value.Value, errors), errors);
    }

    private static double? OptionalNonNegativeNumber(JsonElement json, string name, List<string> errors)
    {
        var value = Field(json, name);
        return value is null ? null : NonNegative(name, Number(name, value.Value, errors), errors);
    }

    private static int? NonNegativeInteger(JsonElement json, string name, List<string> errors)
    {
        var value = Field(json, name);
        if (value is null)
        {
            errors.Add($"{name} を指定してください");
            return null;
        }

        if (value.Value.ValueKind != JsonValueKind.Number || !value.Value.TryGetInt32(out var number))
        {
            errors.Add($"{name} は整数にしてください");
            return null;
        }

        if (number < 0)
        {
            errors.Add($"{name} は 0 以上にしてください");
            return null;
        }

        return number;
    }

    private static double? Number(string name, JsonElement value, List<string> errors)
    {
        if (value.ValueKind != JsonValueKind.Number)
        {
            errors.Add($"{name} は数値にしてください");
            return null;
        }

        return value.GetDouble();
    }

    private static double? NonNegative(string name, double? value, List<string> errors)
    {
        if (value is < 0)
        {
            errors.Add($"{name} は 0 以上にしてください");
            return null;
        }

        return value;
    }

    private static TValue? OneOf<TValue>(
        JsonElement json, string name, IReadOnlyDictionary<string, TValue> choices, List<string> errors)
        where TValue : struct
    {
        var value = Field(json, name);
        if (value is null)
        {
            errors.Add($"{name} を指定してください");
            return null;
        }

        return OptionalOneOf(json, name, choices, errors);
    }

    /// <summary>JSON の値（数値か文字列）を、決められた値の一覧のどれかに対応づける。</summary>
    private static TValue? OptionalOneOf<TValue>(
        JsonElement json, string name, IReadOnlyDictionary<string, TValue> choices, List<string> errors)
        where TValue : struct
    {
        var value = Field(json, name);
        if (value is null)
        {
            return null;
        }

        var raw = value.Value.ValueKind == JsonValueKind.String ? value.Value.GetString() : value.Value.GetRawText();
        if (raw is not null && choices.TryGetValue(raw, out var choice))
        {
            return choice;
        }

        errors.Add($"{name} は {string.Join("、", choices.Keys)} のどれかにしてください");
        return null;
    }
}
